src/rgw/rgw_sal_filter.cc

```diff
@@ -0,0 +1,1362 @@
+{
+  return next->get_cluster_id(dpp, y);
+}
+
+std::unique_ptr<User> FilterStore::get_user(const rgw_user &u)
+{
+  std::unique_ptr<User> user = next->get_user(u);
+  return std::make_unique<FilterUser>(std::move(user));
+}
+
+int FilterStore::get_user_by_access_key(const DoutPrefixProvider* dpp, const std::string& key, optional_yield y, std::unique_ptr<User>* user)
+{
+  std::unique_ptr<User> nu;
+  int ret;
+
+  ret = next->get_user_by_access_key(dpp, key, y, &nu);
+  if (ret != 0)
+    return ret;
+
+  User* u = new FilterUser(std::move(nu));
+  if (!u)
+    return -ENOMEM;
+
+  user->reset(u);
+  return 0;
+}
+
+int FilterStore::get_user_by_email(const DoutPrefixProvider* dpp, const std::string& email, optional_yield y, std::unique_ptr<User>* user)
+{
+  std::unique_ptr<User> nu;
+  int ret;
+
+  ret = next->get_user_by_email(dpp, email, y, &nu);
+  if (ret != 0)
+    return ret;
+
+  User* u = new FilterUser(std::move(nu));
+  if (!u)
+    return -ENOMEM;
+
+  user->reset(u);
+  return 0;
+}
+
+int FilterStore::get_user_by_swift(const DoutPrefixProvider* dpp, const std::string& user_str, optional_yield y, std::unique_ptr<User>* user)
+{
+  std::unique_ptr<User> nu;
+  int ret;
+
+  ret = next->get_user_by_swift(dpp, user_str, y, &nu);
+  if (ret != 0)
+    return ret;
+
+  User* u = new FilterUser(std::move(nu));
+  if (!u)
+    return -ENOMEM;
+
+  user->reset(u);
+  return 0;
+}
+
+std::unique_ptr<Object> FilterStore::get_object(const rgw_obj_key& k)
+{
+  std::unique_ptr<Object> o = next->get_object(k);
+  return std::make_unique<FilterObject>(std::move(o));
+}
+
+int FilterStore::get_bucket(const DoutPrefixProvider* dpp, User* u, const rgw_bucket& b, std::unique_ptr<Bucket>* bucket, optional_yield y)
+{
+  std::unique_ptr<Bucket> nb;
+  int ret;
+  User* nu = nextUser(u);
+
+  ret = next->get_bucket(dpp, nu, b, &nb, y);
+  if (ret != 0)
+    return ret;
+
+  Bucket* fb = new FilterBucket(std::move(nb), u);
+  if (!fb)
+    return -ENOMEM;
+
+  bucket->reset(fb);
+  return 0;
+}
+
+int FilterStore::get_bucket(User* u, const RGWBucketInfo& i, std::unique_ptr<Bucket>* bucket)
+{
+  std::unique_ptr<Bucket> nb;
+  int ret;
+  User* nu = nextUser(u);
+
+  ret = next->get_bucket(nu, i, &nb);
+  if (ret != 0)
+    return ret;
+
+  Bucket* fb = new FilterBucket(std::move(nb), u);
+  if (!fb)
+    return -ENOMEM;
+
+  bucket->reset(fb);
+  return 0;
+}
+
+int FilterStore::get_bucket(const DoutPrefixProvider* dpp, User* u, const std::string& tenant, const std::string& name, std::unique_ptr<Bucket>* bucket, optional_yield y)
+{
+  std::unique_ptr<Bucket> nb;
+  int ret;
+  User* nu = nextUser(u);
+
+  ret = next->get_bucket(dpp, nu, tenant, name, &nb, y);
+  if (ret != 0)
+    return ret;
+
+  Bucket* fb = new FilterBucket(std::move(nb), u);
+  if (!fb)
+    return -ENOMEM;
+
+  bucket->reset(fb);
+  return 0;
+}
+
+bool FilterStore::is_meta_master()
+{
+  return next->is_meta_master();
+}
```

Comment:
for all of these functions that just forward the same call directly to `next`, it might help to move the definitions into `rgw_sal_filter.h` so they're inline. this way if we need to change a sal function signature, we wouldn't need to edit both `rgw_sal_filter.h` and `rgw_sal_filter.cc`

Reply:
I generally put them into rgw_sal_filter.h if they were short.  If they went longer than a line, I put them into the .cc file instead.